tsdb/chunks/head_chunks.go

```diff
@@ -0,0 +1,753 @@
+		newTailFile, err := fileutil.OpenMmapFile(oldFile.Name())
+		if err != nil {
+			return err
+		}
+		cdm.mmappedChunksMtx.Lock()
+		oldMmapFile := cdm.closers[oldSeq]
+		cdm.closers[oldSeq] = newTailFile
+		cdm.mmappedChunkFiles[oldSeq] = realByteSlice(newTailFile.Bytes())
+		cdm.mmappedChunksMtx.Unlock()
+		// Closing the last mmapped file.
+		if err := oldMmapFile.Close(); err != nil {
+			cdm.mmappedChunksMtx.Unlock()
+			return err
+		}
+	}
+
+	mmapFile, err := fileutil.OpenMmapFileWithSize(f.Name(), int(MaxHeadChunkFileSize))
+	if err != nil {
+		return err
+	}
+	cdm.mmappedChunksMtx.Lock()
+	cdm.closers[cdm.curFileSequence] = mmapFile
+	cdm.mmappedChunkFiles[cdm.curFileSequence] = realByteSlice(mmapFile.Bytes())
+	cdm.mmappedChunksMtx.Unlock()
+
+	return nil
+}
+
+// finalizeCurFile writes all pending data to the current tail file,
+// truncates its size, and closes it.
+func (cdm *ChunkDiskMapper) finalizeCurFile() error {
+	if cdm.curFile == nil {
+		return nil
+	}
+
+	if err := cdm.flushBuffer(); err != nil {
+		return err
+	}
+
+	binary.BigEndian.PutUint64(cdm.writeBuf[:], uint64(cdm.curFileMaxt))
+	if _, err := cdm.curFile.WriteAt(cdm.writeBuf[:MintMaxtSize], HeaderMaxtOffset); err != nil {
+		return nil
+	}
+
+	if err := cdm.curFile.Sync(); err != nil {
+		return err
+	}
+
+	return cdm.curFile.Close()
+}
+
+func (cdm *ChunkDiskMapper) write(b []byte) error {
+	n, err := cdm.wbuf.Write(b)
+	atomic.AddInt64(&cdm.n, int64(n))
+	return err
+}
+
+func (cdm *ChunkDiskMapper) writeWithCRC32(b []byte) error {
+	if err := cdm.write(b); err != nil {
+		return err
+	}
+	_, err := cdm.crc32.Write(b)
+	return err
+}
+
+// flushBuffer flushes the current in-memory chunks.
+// Assumes that wbufLock is _write_ locked before calling this method.
+func (cdm *ChunkDiskMapper) flushBuffer() error {
+	if err := cdm.wbuf.Flush(); err != nil {
+		return err
+	}
+	for i := 0; i < inBufferShards; i++ {
+		cdm.inBufferChunksLocks[i].Lock()
+		cdm.inBufferChunks[i] = make(map[uint64]chunkenc.Chunk)
+		cdm.inBufferChunksLocks[i].Unlock()
+	}
+	return nil
+}
+
+// Chunk returns a chunk from a given reference.
+func (cdm *ChunkDiskMapper) Chunk(ref uint64) (chunkenc.Chunk, error) {
+	var (
+		// Get the upper 4 bytes.
+		// These contain the head chunk file index.
+		sgmIndex = int(ref >> 32)
+		// Get the lower 4 bytes.
+		// These contain the head chunk file offset where the encoding for this chunk starts.
+		chkStart = int((ref << 32) >> 32)
+		chkCRC32 = newCRC32()
+	)
+
+	cdm.wbufLock.RLock()
+	if sgmIndex == cdm.curFileSequence && chkStart > int(cdm.n)-cdm.wbuf.Buffered() {
+		chunk := cdm.getChunkFromBuffer(ref)
+		if chunk != nil {
+			cdm.wbufLock.RUnlock()
+			return chunk, nil
+		}
+	}
+	cdm.wbufLock.RUnlock()
+
+	cdm.mmappedChunksMtx.RLock()
+	// We hold this read lock for the entire duration because if the Close()
+	// is called, the data in the byte slice will get corrupted as the mmapped
+	// file will be closed.
```

Comment:
I think that's kind of obvious, I would skip this comment.

Reply:
I would like to keep the comment. Initially, I did not have the lock for the entire code and spent a lot of time debugging to find out why am I getting corrupt data :). So I added the comment after I realized it to not repeat the same mistake in that place.